carma_wm_ctrl/src/WMBroadcaster.cpp

```diff
@@ -197,104 +296,330 @@ std::shared_ptr<Geofence> WMBroadcaster::geofenceFromMsg(const cav_msgs::Traffic
+  bool detected_workzone_signal = msg_v01.package.label_exists && msg_v01.package.label.find("SIG_WZ") != std::string::npos;
+
+  if (detected_workzone_signal) // if workzone message detected, save to cache to process later
+  {
+    gf_ptr->label_ = msg_v01.package.label; // to extract intersection, and signal group id
+    work_zone_geofence_cache_[msg_detail.choice] = gf_ptr;
+    if (work_zone_geofence_cache_.size() < 4)
+    {
+      ROS_INFO_STREAM("Received 'SIG_WZ' signal. Waiting for the rest of the messages, returning for now...");
+      return {};
+    }
+    else
+    {
+      return_list = createWorkzoneGeofence(work_zone_geofence_cache_);
+      return return_list;
+    }
+  }
+  else if (msg_detail.choice == cav_msgs::TrafficControlDetail::CLOSED_CHOICE && msg_detail.closed==cav_msgs::TrafficControlDetail::CLOSED) // if stand-alone closed signal apart from Workzone
+  {
+    addRegionAccessRule(gf_ptr,msg_v01,affected_llts);
+  }
+ 
+  return_list.push_back(gf_ptr);
+  return return_list;
+}
+
+std::vector<std::shared_ptr<Geofence>> WMBroadcaster::createWorkzoneGeofence(std::unordered_map<uint8_t, std::shared_ptr<Geofence>> work_zone_geofence_cache)
+{
+  std::shared_ptr<std::vector<lanelet::Lanelet>> parallel_llts = std::make_shared<std::vector<lanelet::Lanelet>>(std::vector<lanelet::Lanelet>());
+  std::shared_ptr<std::vector<lanelet::Lanelet>> middle_opposite_lanelets = std::make_shared<std::vector<lanelet::Lanelet>>(std::vector<lanelet::Lanelet>());
   
-  ros::Time end_time = msg_schedule.end;
-  if (!msg_schedule.end_exists) {
-    ROS_DEBUG_STREAM("No end time for geofence, using ros::TIME_MAX");
-    end_time = ros::TIME_MAX; // If there is no end time use the max time
+  // Split existing lanelets and filter into parallel_llts and middle_opposite_lanelets
+  preprocessWorkzoneGeometry(work_zone_geofence_cache, parallel_llts, middle_opposite_lanelets);
+
+  // Create geofence and rest of the required lanelets along with traffic light and stop rule for completing workzone area
+  auto gf_ptr = createWorkzoneGeometry(work_zone_geofence_cache, parallel_llts->front(), parallel_llts->back(), middle_opposite_lanelets );
+
+  // copy static info from the existing workzone
+  gf_ptr->id_ = work_zone_geofence_cache[cav_msgs::TrafficControlDetail::TAPERRIGHT]->id_; //using taperright's id as the whole geofence's id
+  // schedule
+  gf_ptr->schedules = work_zone_geofence_cache[cav_msgs::TrafficControlDetail::TAPERRIGHT]->schedules; //using taperright's schedule as the whole geofence's schedule
+  return {gf_ptr};
+}
+
+std::shared_ptr<Geofence> WMBroadcaster::createWorkzoneGeometry(std::unordered_map<uint8_t, std::shared_ptr<Geofence>> work_zone_geofence_cache,  lanelet::Lanelet parallel_llt_front,  lanelet::Lanelet parallel_llt_back, 
+                                                    std::shared_ptr<std::vector<lanelet::Lanelet>> middle_opposite_lanelets)
+{
+  auto gf_ptr = std::make_shared<Geofence>(Geofence());
```

Comment:
Just so you know you can just do `std::make_shared<Geofence>()`

Reply:
The make_shared call actually takes in the constructor arguments directly. Its a pretty wild method

Reply:
awesome! replaced all instances